[Divider][material-next] Add Divider component

Best-Sardar opened the pull request:
<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->

- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/material-ui/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

related to: #39007

mj12albert commented:
@Best-Sardar Thanks for working on this ✨ 

I've cherry-picked your first commit in a separate [PR](https://github.com/mui/material-ui/pull/39197) to make the diffs in this one easier to read ~ CC @DiegoAndai 

Would you mind rebasing once we merge that one into master 🙏

mj12albert commented:
@Best-Sardar Actually there's only a very minor conflict 😅

I cloned your branch to test resolving it, and `git rebase upstream/master -X ours` worked for me, though you could do it however you like ~

Best-Sardar commented:
Hello @mj12albert,
This is my first experience working on open-source projects, which is why I have made an effort to submit all the changes in the form of one pull request (PR).
I have also created separate commits to make the review process easier. By reviewing each commit individually, we can avoid creating multiple PRs.

I have completed the rebase, and I think we are now in the correct position.
I am actively searching for any pipeline problems and working to resolve them.

It would greatly please me if you could point out any issues that would help me progress in the field of working on open-source projects.

Thanks

mj12albert commented:
@Best-Sardar Hey looks like most of the checks are passing 👍 

For the remaining:
- `test_static`: run `yarn proptypes` and commit the changes to update the PropTypes generated from TS types
- `test-dev`: this is happening on other PRs too but it's somewhat intermittent, I just retried it

DiegoAndai commented on packages/mui-material-next/src/Divider/Divider.tsx:
```diff
@@ -54,48 +55,52 @@ const DividerRoot = styled('div', {
+      borderWidth: 0,
+      borderStyle: 'solid',
+      borderColor: tokens.sys.color.outlineVariant,
+      borderBottomWidth: 'thin',
```
I think we should use `1px` instead of `thin` (here and on the other places `thin` is used), as that's the spec from Material You

DiegoAndai commented on packages/mui-material-next/src/Divider/Divider.tsx:
```diff
@@ -54,48 +55,52 @@ const DividerRoot = styled('div', {
-      ownerState.orientation === 'vertical' && {
-        marginTop: theme.spacing(1),
-        marginBottom: theme.spacing(1),
+      ...(ownerState.light && {
+        borderColor: theme.vars
+          ? `rgba(${theme.vars.palette.dividerChannel} / 0.08)`
+          : alpha(theme.palette.divider, 0.08),
       }),
```
How might we support the [`light` prop](https://mui.com/material-ui/api/divider/#Divider-prop-light) within Material You? Should we drop it? `outlineVariant` is already the lighter outline token.

@zanivan

zanivan replied:
By the [MY guidelines](https://m3.material.io/components/divider/specs#34f44feb-4251-457b-9a35-6ab77fbbaf6a), by default, the divider uses the `outline-variant`—that's the `palette.neutral-variant80` for the light theme and `palette.neutral-variant30` for the dark theme. 

As I see, if we're using our `outlineVariant`, it's currently accurate with the guidelines from MY, even with it as the lightest color we have. All in all, I think we can either drop the `light` prop, or if this ends up being a problem, we could change it to be `true` as default and have a darker tone for when it's set to `false`.

DiegoAndai replied:
So the options would be

1. Drop the `light` prop and use `outlineVariant`. If users want to use `outline` or another color, they can customize it. We can add a `color` prop if necessary, which is more in line with other components.
2. Keep the `light` prop but change the default to `true`. The Divider would then use `outlineVariant` by default and `outline` when `light={false}`. The default change would be a breaking change.
3. Keep the `light` prop with the same default. The Divider would then use `outline` by default and `outlineVariant` when `light={true}`. This would mean the Divider won't follow the MY guideline by default.

I lean towards 1. because:
- Using customization and/or a `color` prop makes more sense with the rest of the components and is more flexible.
- Option 2. is a middle ground, but the default being `true` would be a weird API. Also, it's still a breaking change as users would need to address if they're using the `light` prop to differentiate two styles of Dividers.
- Option 3. is not desirable, as we should follow the MY guidelines by default as much as possible.

What do you think @mnajdova?

DiegoAndai replied:
Hey @Best-Sardar, let's go with option 1: Drop the `light` prop. We need to add this to the `migration.md` file. We need to explain to users how to achieve the same through style overrides or the sx prop in the migration file. As a guideline, you can follow: https://github.com/mui/material-ui/blob/master/packages/mui-material-next/migration.md#chip

In this case, I would have an example with the [sx prop](https://mui.com/material-ui/customization/how-to-customize/#the-sx-prop).

Let me know if you need any help with that

DiegoAndai commented on packages/mui-material-next/src/Divider/Divider.tsx:
```diff
@@ -54,48 +55,52 @@ const DividerRoot = styled('div', {
+      }),
+      ...(ownerState.variant === 'middle' &&
+        ownerState.orientation === 'horizontal' && {
+          marginLeft: theme.spacing(2),
+          marginRight: theme.spacing(2),
+        }),
+      ...(ownerState.variant === 'middle' &&
+        ownerState.orientation === 'vertical' && {
+          marginTop: theme.spacing(1),
+          marginBottom: theme.spacing(1),
```
Material You specs specify 16dp (16px) for the middle inset variants. Let's use `16px` directly instead of the theme spacing.

@mnajdova I think we still need to figure out how theme spacing (and similar concepts) will work alongside Material You tokens. Do you think we should focus on this soon? Or has this already been discussed?

Best-Sardar replied:
I updated the values based on the Figma file created by MD3 and hardcoded them (set 16px). If any changes are needed, please let me know so they can be applied.

DiegoAndai commented on packages/mui-material-next/src/Divider/Divider.tsx:
```diff
@@ -20,8 +21,8 @@ const useUtilityClasses = (ownerState) => {
       flexItem && 'flexItem',
-      children && 'withChildren',
-      children && orientation === 'vertical' && 'withChildrenVertical',
+      // TODO: Check next line with maintainers
+      ...(children ? ['withChildren', orientation === 'vertical' && 'withChildrenVertical'] : []),
```
We should remove the `withChildrenVertical` composed class, so this line should become only

```suggestion
     children && 'withChildren',
```

This is because we're [dropping composed classes in v6](https://github.com/mui/material-ui/blob/master/packages/mui-material-next/migration.md#remove-composed-css-classes-and-styleoverrides-keys). We should add this to the migration guide, similar to [this example](https://github.com/mui/material-ui/blob/master/packages/mui-material-next/migration.md#removed-combined-anchor-overlap-styleoverrides-keys).

Best-Sardar replied:
Due to the type defined in composeClasses and your request to leave other files untouched, I need to define this line as follows: 
`...(children ? ['withChildren'] : []) `
Is this acceptable?

DiegoAndai replied:
`children && 'withChildren'` should work, right?

Best-Sardar replied:
When the condition is written as `children && 'withChildren'` , considering the "children" types, a situation is created in which "0" is placed in the "root" types, leading to an error. The error message is as follows:

![image](https://github.com/mui/material-ui/assets/21319493/2204ae7d-4d76-4e2f-b11f-1bfb89540bea)

DiegoAndai replied:
Oh, what about `!!children && 'withChildren'` 😊

DiegoAndai reviewed (commented):
Hey, thanks for working on this!

I left a couple of questions 😊 

There are some changes that are not related to the Divider component that we should move out of this PR, so we can keep it as small as possible:
- color scheme changes
- getThemeProps changes

I also tagged @zanivan so he can help us with the review from the design perspective. As well as @mnajdova for a quick question. On those comments, let's wait for what they reply.

Let me know if anything isn't clear or you need help with.

Best-Sardar commented on packages/mui-material-next/src/Divider/Divider.tsx:
```diff
@@ -219,7 +230,7 @@ const Divider = React.forwardRef(function Divider(inProps, ref) {
  * does not get highlight/focus inside of MUI List.
  */
-Divider.muiSkipListHighlight = true;
+Object.assign(Divider, { muiSkipListHighlight: true });
```
Yes, we had a Typescript error. This value is used as metadata, and typing it is difficult due to the presence of react.forwardRef.

Considering the parts that used this data (which were not type-safe), I decided to address this issue by using Object.assign.

DiegoAndai replied:
I think we can do something similar to [Joy's Divider](https://github.com/mui/material-ui/blob/master/packages/mui-joy/src/Divider/Divider.tsx#L209) and ignore it as it's an internal flag

```suggestion
// @ts-ignore internal logic
Divider.muiSkipListHighlight = true;
```

Best-Sardar commented:
Hi @DiegoAndai,
I rebased my branch from master and addressed some of the issues.
By running "yarn proptyps," new changes were added. Are the added values correct?
Is there any documentation available for that script so we can understand its functionality?

Best-Sardar commented:
Dear @DiegoAndai,
Thank you for taking the time to review. I have addressed the raised issues, and there are two remaining items that require @mnajdova's decision.

In the conducted review, are all files under consideration? Specifically, I am referring to the "docs" section. Do we need any changes in that section?

DiegoAndai commented:
Hey @Best-Sardar, I provided a resolution to one of the unresolved comments so you can implement it (remove `light` prop). The other unresolved question I asked @mnajdova is not a blocker for this PR. So let's remove the `light` prop, add it to the migration, and I can review again 😊

DiegoAndai commented on packages/mui-material-next/migration.md:
```diff
@@ -337,3 +337,18 @@ If you are using the `thumb` or `valueLabel` Slider slots, then make sure the co
+
+## Divider
+
+### Removed the "light" prop
```
Let's add in this section that because the "light" prop was removed, the "light" class is also removed. We can change the title to "Removed the "light" prop and class"

DiegoAndai commented on packages/mui-material-next/migration.md:
```diff
@@ -337,3 +337,18 @@ If you are using the `thumb` or `valueLabel` Slider slots, then make sure the co
 
 This is required in v6 as it's used to apply the overlap styles to these slots. For more info take a look into [Material You's Slider overlapping handles guidelines](https://m3.material.io/components/sliders/guidelines#ad5ceb95-a690-4ddd-8243-53a8e13bdab6).
+
+## Divider
```
Let's add a section about the composed class `"withChildrenVertical"` being removed, and how to combine `"withChildren"` and `"vertical"` to achieve the same result.

DiegoAndai commented on packages/mui-material-next/src/Divider/Divider.tsx:
```diff
@@ -54,48 +51,52 @@ const DividerRoot = styled('div', {
-      ownerState.orientation === 'vertical' && {
-        marginTop: theme.spacing(1),
-        marginBottom: theme.spacing(1),
+      ...(ownerState.variant === 'inset' &&
+        ownerState.orientation === 'horizontal' && {
+          marginLeft: '16px',
+        }),
+      ...(ownerState.variant === 'inset' &&
+        ownerState.orientation === 'vertical' && {
+          marginTop: '16px',
+        }),
```
Something is not clear to me regarding the inset prop specs from Material You:

They mention `16px` for the inset, but I think that might be an error because from the [What's new section](https://m3.material.io/components/divider/overview#a7783158-c068-4f06-816d-c90922d404ed), looks like inset dividers have way more than 16px margin.

From Figma, the margin also seems to be `16px`. I think we should go with `16px`, but I just wanted to discuss it as it seems weird from Material's specs side.

What do you think @Best-Sardar @zanivan?

zanivan replied:
I think we should keep `16px` as in the MY docs. 

In the example from _What's new section_ (above), for me, it seems that the divider is actually full width on the list content, and the avatar is aside.

![image](https://github.com/mui/material-ui/assets/37222944/711d74a0-bb05-4aa5-a74b-245a2368954d)

In the majority of examples, it seems to respect the inset margin of `16px`:

<img width="250" alt="Screenshot 2023-10-13 at 14 39 15" src="https://github.com/mui/material-ui/assets/37222944/0a0fe787-bc4e-450c-890f-542ded2038f9">

<img width="250" alt="Screenshot 2023-10-13 at 14 39 15" src="https://github.com/mui/material-ui/assets/37222944/bfdd8a11-2dd8-4680-877f-b98ce2fed821">

<img width="250" alt="Screenshot 2023-10-13 at 14 39 15" src="https://github.com/mui/material-ui/assets/37222944/87ba4753-9ac2-41e4-b314-e6bdd444b439">

DiegoAndai replied:
Makes sense, thanks @zanivan 😊

DiegoAndai commented on docs/data/material/components/dividers/DividerMaterialYouPlayground.js:
```diff
@@ -0,0 +1,101 @@
+          defaultValue: 'horizontal',
+          onChange: (e) => setOrientation(e.target.value),
+        },
+        {
+          propName: 'showLabel',
+          knob: 'switch',
+          defaultValue: false,
+          onChange: () => {
+            setShowLabel((prev) => !prev);
+          },
+        },
```
I think we should remove this one because it might be confusing, as `showLabel` is not a Divider prop. We don't want to mislead people.

DiegoAndai reviewed (commented):
Hey, nice work! I think we're almost there, just a couple of comments and suggestions

DiegoAndai reviewed (approved):
Nice work! I'll checkout the failing builds and merge as soon as possible.

I will review the [Snackbar PR](https://github.com/mui/material-ui/pull/39232#issuecomment-1750428954) as well so we can continue with that one.
